FlowCrypt/src/main/java/com/flowcrypt/email/service/ProcessingOutgoingMessageInfoHelper.kt

```diff
@@ -0,0 +1,386 @@
+      e.printStackTrace()
+      ExceptionUtil.handleError(ForceHandlingException(e))
+
+      var msgEntity = roomDatabase.msgDao().getMsg(email, label, uid)
+        ?: MessageEntity.genMsgEntity(email, label, uid, outgoingMsgInfo)
+
+      if (newMsgId <= 0) {
+        newMsgId = roomDatabase.msgDao().insert(msgEntity)
+        msgEntity = msgEntity.copy(id = newMsgId)
+      }
+
+      if (newMsgId > 0) {
+        when (e) {
+          is NoKeyAvailableException -> {
+            roomDatabase.msgDao().update(
+              msgEntity.copy(
+                state = MessageState.ERROR_PRIVATE_KEY_NOT_FOUND.value,
+                errorMsg = if (TextUtils.isEmpty(e.alias)) e.email else e.alias
+              )
+            )
+          }
+
+          else -> {
+            roomDatabase.msgDao().update(
+              msgEntity.copy(
+                state = MessageState.ERROR_DURING_CREATION.value,
+                errorMsg = e.message
+              )
+            )
+          }
+        }
+      } else {
+        ExceptionUtil.handleError(IllegalStateException("An error occurred during inserting a new message"))
+      }
+
+      val failedOutgoingMsgsCount =
+        roomDatabase.msgDao().getFailedOutgoingMsgsCount(accountEntity.email)
+      if (failedOutgoingMsgsCount > 0) {
+        ErrorNotificationManager(context).notifyUserAboutProblemWithOutgoingMsgs(
+          accountEntity,
+          failedOutgoingMsgsCount
+        )
+      }
+    }
+
+    if (newMsgId > 0) {
+      updateOutgoingMsgCount(email, accountEntity.accountType, roomDatabase)
+    }
+  }
+
+  private fun getAttsCacheDir(context: Context): File {
+    val attsCacheDir = File(context.cacheDir, Constants.ATTACHMENTS_CACHE_DIR)
+    if (!attsCacheDir.exists()) {
+      if (!attsCacheDir.mkdirs()) {
+        throw IllegalStateException("Create cache directory " + attsCacheDir.name + " filed!")
+      }
+    }
+    return attsCacheDir
+  }
+
+  private fun updateOutgoingMsgCount(
+    email: String,
+    accountType: String?,
+    roomDatabase: FlowCryptRoomDatabase
+  ) {
+    val outgoingMsgCount = roomDatabase.msgDao().getOutboxMsgs(email).size
+    val outboxLabel =
+      roomDatabase.labelDao().getLabel(email, accountType, JavaEmailConstants.FOLDER_OUTBOX)
+
+    outboxLabel?.let {
+      roomDatabase.labelDao().update(it.copy(messagesTotal = outgoingMsgCount))
+    }
+  }
+
+  private fun prepareMessageEntity(
+    accountEntity: AccountEntity,
+    msgInfo: OutgoingMessageInfo,
+    generatedUID: Long,
+    msg: Message,
+    rawMsg: String,
+    attsCacheDir: File
+  ): MessageEntity {
+
+    val messageEntity = MessageEntity.genMsgEntity(
+      accountEntity.email,
+      JavaEmailConstants.FOLDER_OUTBOX, msg, generatedUID, false
+    )
+
+    val hasAtts = msgInfo.atts?.isNotEmpty() == true || msgInfo.forwardedAtts?.isNotEmpty() == true
+    val isEncrypted = msgInfo.encryptionType === MessageEncryptionType.ENCRYPTED
+    val msgStateValue =
```

Comment:
This one could be `msgState` and then use it as `msgState.value` - to me, such code is a bit shorter and a bit more clear